### PR TITLE
#1788 Verification if module source point to existing directory

### PR DESCRIPTION
**Pull request by denis256**
Updated source downloading logic to verify if module directory actually exists

Example:
```

$ cat terragrunt.hcl
terraform {
  source = "git::ssh://git@github.com/denis256/internal-terragrunt-test-cases.git//module-666/invalid-path?ref=main"
}

$ terragrunt init
ERRO[0001] Working dir module-666/invalid-path from source git::ssh://git@github.com/denis256/internal-terragrunt-test-cases.git?ref=main does not exist 
ERRO[0001] Unable to determine underlying exit code, so Terragrunt will exit with error code 1 


```

Included changes:
 * Added verification of terraformSource.WorkingDir after downloading of sources
 * Added test cases

Fix for issue: #1788

### CODE REVIEW

File: cli/download_source.go [hunk 1]
```diff
@@ -158,3 +165,32 @@ func downloadSource(terraformSource *tfsource.TerraformSource, terragruntOptions
...
+// Check if working terraformSource.WorkingDir exists and is directory
+func validateWorkingDir(terraformSource *tfsource.TerraformSource) error {
+	fileInfo, err := os.Stat(terraformSource.WorkingDir)
+	if err != nil {
+		return WorkingDirNotFound{terraformSource.WorkingDir}
+	}
+
+	if !fileInfo.IsDir() {
+		return WorkingDirNotDir{terraformSource.WorkingDir}
+	}
```
Comment (brikis98) [inline cli/download_source.go:178]: NIT: could use `util.IsDir`

File: cli/download_source.go [hunk 2]
```diff
@@ -158,3 +165,32 @@ func downloadSource(terraformSource *tfsource.TerraformSource, terragruntOptions
...
+	return nil
+}
+
+type WorkingDirNotFound struct {
+	Dir string
+}
+
+func (err WorkingDirNotFound) Error() string {
+	return fmt.Sprintf("Directory not found: %s", err.Dir)
```
Comment (brikis98) [inline cli/download_source.go:187]: We may want a more specific error message that includes the `source` URL. E.g., "Working dir XXX from source URL YYY does not exist."

**Comment by brikis98** [comment]
Thanks! I'll wait for https://github.com/gruntwork-io/terragrunt/pull/1799 to merge before kicking off tests.

**Comment by brikis98** [comment]
OK, https://github.com/gruntwork-io/terragrunt/pull/1799 is now merged. Could you pull in the latest from `master`?

**Comment by denis256** [comment]
Updated with changes from  #1799

**Comment by brikis98** [comment]
Great, thanks! Re-running tests now.

**Comment by brikis98** [comment]
Looks like a pre-commit check failed:

```
goimports................................................................Failed
- hook id: goimports
- files were modified by this hook

cli/download_source.go
cli/download_source_test.go
```

I think you need to run `goimports`.

**Comment by denis256** [comment]
Fixed, will try to make GoLand do `goimports` automatically

**Review by brikis98 — APPROVED** [review]
Thanks! Kicking off tests once more.

**Comment by brikis98** [comment]
Tests passed! Merging now.

**Comment by brikis98** [comment]
https://github.com/gruntwork-io/terragrunt/releases/tag/v0.31.10

**Pull request merged**